packages/core/src/model-types.ts

```diff
@@ -0,0 +1,52 @@
+type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>);
```

Comment:
It was a pretty interesting finding for me as well.😄 Check this issue for more technical details -> https://github.com/microsoft/TypeScript/issues/29729

A simple explanation would be:

If union literal strings with `string`, the type system will collapse them all together into `string`. But the beauty of this trick is that it makes the type system think `string & Record<never, never>` is not `string`, and thus, not collapsible with literal string, so that literal strings get to survive.